From 92ddec026f65fe6fd9e389914f1c5bde4ba28667 Mon Sep 17 00:00:00 2001
From: thampan <joby.thampan@intel.com>
Date: Fri, 19 Jul 2019 19:39:06 +0800
Subject: [PATCH] PONRTSYS-4542: 2nd Multicast Entry with same hash index as
 the first entry, HW was not able to search and find. Software was not adapted
 to the B0 Hardware modification. First Index in Software Multicast Table MSB
 bit cannot be set to 1, otherwise HW will stop search.

---
 drivers/net/ethernet/lantiq/switch-api/gsw_swmcast.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_swmcast.c b/drivers/net/ethernet/lantiq/switch-api/gsw_swmcast.c
index f61999fc05dc..c626d322e770 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_swmcast.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_swmcast.c
@@ -313,7 +313,14 @@ static int set_pce_hash_table(void *cdev, MCAST_HASHTBL *phtable, u32 loc)
 	pcetable.pcindex = loc;
 	pcetable.table = PCE_MULTICAST_SW_INDEX;
 
-	pcetable.key[0] = phtable->first_idx;
+	/* Cannot add 0x3FF to First Index, Msb Bit is used
+	 * for stopping search
+	 */
+	if (phtable->first_idx == 0xFFFF)
+		pcetable.key[0] = 0;
+	else
+		pcetable.key[0] = phtable->first_idx;
+	
 	pcetable.key[1] = phtable->nxt_idx;
 	pcetable.valid = phtable->valid;
 	pcetable.key[2] = ((phtable->src_ip_mode & 0x3) << 14);
